official/benchmark/benchmark_uploader.py

```diff
@@ -95,39 +82,54 @@ def upload_metric(self, dataset_name, table_name, run_id):
         benchmark_run table.
       run_id: string, a unique ID that will be attached to the data, usually
         this is a UUID4 format. This should be the same as the benchmark run_id.
+      metric_json_list: list, a list of JSON object that record the metric info.
+    """
+    for m in metric_json_list:
+      m["run_id"] = run_id
+    table_ref = self._bq_client.dataset(dataset_name).table(table_name)
+    errors = self._bq_client.insert_rows_json(table_ref, metric_json_list)
+    if errors:
+      tf.logging.error(
+          "Failed to upload benchmark info to bigquery: {}".format(errors))
+
+  def upload_benchmark_run_file(
+      self, dataset_name, table_name, run_id, run_json_file):
+    """Upload benchmark run information to Bigquery from input json file.
+
+    Args:
+      dataset_name: string, the name of bigquery dataset where the data will be
+        uploaded.
+      table_name: string, the name of bigquery table under the dataset where
+        the data will be uploaded.
+      run_id: string, a unique ID that will be attached to the data, usually
+        this is a UUID4 format.
+      run_json_file: string, the file path that contains the run JSON data.
+    """
+    with tf.gfile.GFile(run_json_file) as f:
+      benchmark_json = json.load(f)
+      self.upload_benchmark_run_json(
+          dataset_name, table_name, run_id, benchmark_json)
+
+  def upload_metric_file(
+      self, dataset_name, table_name, run_id, metric_json_file):
+    """Upload metric information to Bigquery from input json file.
+
+    Args:
+      dataset_name: string, the name of bigquery dataset where the data will be
+        uploaded.
+      table_name: string, the name of bigquery table under the dataset where
+        the metric data will be uploaded. This is different from the
+        benchmark_run table.
+      run_id: string, a unique ID that will be attached to the data, usually
+        this is a UUID4 format. This should be the same as the benchmark run_id.
+      metric_json_file: string, the file path that contains the metric JSON
+        data.
     """
-    expected_file = os.path.join(
-        self._logging_dir, logger.METRIC_LOG_FILE_NAME)
-    with tf.gfile.GFile(expected_file) as f:
+    with tf.gfile.GFile(metric_json_file) as f:
       lines = f.readlines()
```

Comment:
It's easy to imagine this consuming all of the memory for a really big file. Why not do this as a generator? (That is, for line in f.read(), or whatever the correct syntax is, rather than reading in the whole file upfront.)

Reply:
The side effect of that is we will have one API call per line. I could implement some buffering and limit the max number of lines we load for each upload, if this is what u want. At the moment, we don't see a memory issue since usually we only have O(100) lines of metric record for small model. For big model like imagenet, it will be covered by the streaming upload.

Reply:
But isn't this exactly the same as:

```
for line in f:
  if line.strip():
    metrics.append(json.loads(line))
```
Except that readlines requires all lines read into memory, and the above creates a generator that reads line by line?

Reply:
Done. Thanks for the suggestion.